### PR TITLE
Friends List

### PR DESCRIPTION
**Pull request by sbaum23**
Adds the ability to add/remove friends as well as see the list of friends when you log on.  The status is also accurate and will change as friends log on/off.  I added a Logout method to the Player class for putting stuff in there that needs to be done when a player logs off (like sending a packet to all your online friends that you are offline).

The extra commits are due to rebasing.

This resolves issue #39.

### CODE REVIEW

File: Source/ACE.Database/CharacterDatabase.cs [hunk 1]
```diff
@@ -266,7 +274,21 @@ public Task UpdateCharacter(Character character)
...
+                    f.Id = new ObjectGuid(result.Read<uint>(i, "friendId"), GuidType.Player);
+                    f.Name = result.Read<string>(i, "name");
+
+                    // Not sure we want the actually load these next two values.  They are passed in the packet, but their purpose is unknown.
+                    f.FriendIdList = new List<ObjectGuid>();
+                    f.FriendOfIdList = new List<ObjectGuid>();
+                    c.AddFriend(f);
+                }
+
+                LoadCharacterProperties(c);
```
Comment (LtRipley36706) [inline Source/ACE.Database/CharacterDatabase.cs]: Change this line to: await LoadCharacterProperties(c);

File: Source/ACE.Entity/Character.cs [hunk 2]
```diff
@@ -178,6 +181,27 @@ private void AddSkill(Skill skill, SkillStatus status, uint ranks)
         }
 
         /// <summary>
+        /// Returns a read only list of the characters friends
+        /// </summary>
+        public ReadOnlyCollection<Friend> GetFriends()
+        {
+            return new ReadOnlyCollection<Friend>(friends);
```
Comment (Mogwai-AC) [inline Source/ACE.Entity/Character.cs]: This is going to cause the full recreation of the collection every time the property is hit.  You need to cache the read-only collection and reset it any time the base friends collection is reassigned.

File: Source/ACE/Entity/Player.cs [hunk 3]
```diff
@@ -339,6 +345,84 @@ private uint SpendSkillXp(CharacterSkill skill, uint amount)
...
+        }
+
+        public async Task<string> AddFriend(string friendName)
+        {
+            if (string.Equals(friendName, Name, StringComparison.CurrentCultureIgnoreCase))
+                return "Sorry, but you can't be friends with yourself.";
+
+            // Check if friend exists
+            if (character.GetFriends().SingleOrDefault(f => string.Equals(f.Name, friendName, StringComparison.CurrentCultureIgnoreCase)) != null)
+            return "That character is already in your friends list";
```
Comment (Mogwai-AC) [inline Source/ACE/Entity/Player.cs]: indentation

File: Source/ACE/Network/GameEvent/Events/GameEventFriendsListUpdate.cs [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+        protected override void WriteEventBody()
+        {
+            List<Friend> friendList = null;
+
+            if (updateType == FriendsUpdateTypeFlag.FullList)
+                friendList = session.Player.GetFriends().ToList();
+            else
+                friendList = new List<Friend>() { friend };
+
+            fragment.Payload.Write(friendList.Count);
```
Comment (Mogwai-AC) [inline Source/ACE/Network/GameEvent/Events/GameEventFriendsListUpdate.cs]: count is int.  this is probably expecting a uint.  please verify.

File: Source/ACE/Network/GameEvent/Events/GameEventFriendsListUpdate.cs [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+                    isOnline = onlineStatusVal;
+                else if (friendSession != null && friendSession.Player.IsOnline)
+                    isOnline = true;
+
+                fragment.Payload.Write(f.Id.Full); // friend Object ID
+                fragment.Payload.Write(isOnline ? 1u : 0u); // is Online               
+                fragment.Payload.Write(0u); // Unknown
+                fragment.Payload.WriteString16L(f.Name); // Friend Name
+
+                fragment.Payload.Write(f.FriendIdList.Count); // Number of people on this persons friend's list.
```
Comment (Mogwai-AC) [inline Source/ACE/Network/GameEvent/Events/GameEventFriendsListUpdate.cs]: List.Count is int, this is probably supposed to be uint.  Please verify.

File: Source/ACE/Network/GameEvent/Events/GameEventFriendsListUpdate.cs [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+                foreach (var fid in f.FriendIdList)
+                    fragment.Payload.Write(fid.Full);
+
+                fragment.Payload.Write(f.FriendOfIdList.Count); // Number of people that have this person as a friend.
```
Comment (Mogwai-AC) [inline Source/ACE/Network/GameEvent/Events/GameEventFriendsListUpdate.cs]: List.Count is int, this is probably supposed to be uint.  Please verify.

**Pull request closed without merging**